from fastapi import HTTPException
from sqlalchemy.orm import Session
import bcrypt
import models, schemas

def get_user(db: Session, user_id: int):
    return db.query(models.User).filter(models.User.id == user_id).first()

def get_user_by_username(db: Session, username: str):
    return db.query(models.User).filter(models.User.username == username).first()

def create_user(db: Session, user: schemas.UserCreate):
    hashed_pw = hash_password(user.password)
    db_user = models.User(
        username=user.username,
        password=hashed_pw
    )
    db.add(db_user)
    db.commit()
    db.refresh(db_user)
    return db_user

def login_user(db: Session, username: str, password: str) -> schemas.User:
    user = db.query(models.User).filter_by(username=username).first()
    if not user:
        raise HTTPException(
            status_code=401, detail="Invalid username or password"
        )
    if not verify_password(password, user.password):
        raise HTTPException(
            status_code=401, detail="Invalid username or password"
        )
    return user

def hash_password(password: str) -> str:
    password_bytes = password.encode('utf-8')
    hashed_password = bcrypt.hashpw(password_bytes, bcrypt.gensalt())
    return hashed_password.decode('utf-8')

def verify_password(plain_password: str, hashed_password: str) -> bool:
    return bcrypt.checkpw(plain_password.encode('utf-8'), hashed_password.encode('utf-8'))
